crates/ghost_actor/src/lib.rs

```diff
@@ -87,116 +96,100 @@ mod tests {
             Self {
-                actor_state: Some(GhostActorState::new()),
+                channel_parent: Some(channel_parent),
+                channel_self: Some(channel_self.as_context_channel()),
             }
         }
     }
 
     impl
         GhostActor<
-            i8,
             dht_protocol::RequestToParent,
             dht_protocol::RequestToParentResponse,
             dht_protocol::RequestToChild,
             dht_protocol::RequestToChildResponse,
-            String,
+            FakeError,
         > for RrDht
     {
         fn as_any(&mut self) -> &mut dyn Any {
             &mut *self
         }
 
-        fn get_actor_state(
-            &mut self,
-        ) -> &mut GhostActorState<
-            i8,
-            dht_protocol::RequestToParent,
-            dht_protocol::RequestToParentResponse,
-            dht_protocol::RequestToChildResponse,
-            String,
-        > {
-            self.actor_state.as_mut().unwrap()
-        }
-
-        fn take_actor_state(
-            &mut self,
-        ) -> GhostActorState<
-            i8,
-            dht_protocol::RequestToParent,
-            dht_protocol::RequestToParentResponse,
-            dht_protocol::RequestToChildResponse,
-            String,
-        > {
-            std::mem::replace(&mut self.actor_state, None).unwrap()
-        }
-
-        fn put_actor_state(
+        fn take_parent_channel(
             &mut self,
-            actor_state: GhostActorState<
-                i8,
+        ) -> Option<
+            GhostChannel<
+                dht_protocol::RequestToChild,
+                dht_protocol::RequestToChildResponse,
                 dht_protocol::RequestToParent,
                 dht_protocol::RequestToParentResponse,
-                dht_protocol::RequestToChildResponse,
-                String,
+                FakeError,
             >,
-        ) {
-            std::mem::replace(&mut self.actor_state, Some(actor_state));
+        > {
+            std::mem::replace(&mut self.channel_parent, None)
         }
 
-        // our parent is making a request of us
-        fn request(
-            &mut self,
-            request_id: Option<RequestId>,
-            request: dht_protocol::RequestToChild,
-        ) {
-            match request {
-                dht_protocol::RequestToChild::ResolveAddressForId { id } => {
-                    println!("dht got ResolveAddressForId {}", id);
-                    if let Some(request_id) = request_id {
-                        println!("dht ResolveAddressForId responding to parent");
-                        self.get_actor_state().respond_to_parent(
-                            request_id,
-                            dht_protocol::RequestToChildResponse::ResolveAddressForId(Ok(
+        fn process_concrete(&mut self) -> GhostResult<WorkWasDone> {
+            let mut channel_self = std::mem::replace(&mut self.channel_self, None);
```

Comment:
this pattern looks totally ripe for the same kind of macro as I saw you did for `ghost_cb_call`...